jwst/extract_1d/ifu.py

```diff
@@ -384,14 +413,107 @@ def extract_ifu(input_model, source_type, extract_params):
     return (ra, dec, wavelength, net, background, npixels, dq)
 
 
-def image_extract_ifu(input_model, extract_params):
+def locn_from_wcs(input_model, ra_targ, dec_targ):
+    """Get the location of the spectrum, based on the WCS.
+
+    Parameters
+    ----------
+    input_model : data model
+        The input science model.
+
+    ra_targ, dec_targ : float or None
+        The right ascension and declination of the target (degrees).
+
+    Returns
+    -------
+    locn : tuple of two int, or None
+        X and Y coordinates (in that order) of the pixel that has right
+        ascension and declination coordinates closest to the target
+        location.  The spectral extraction region should be centered here.
+    """
+
+    if ra_targ is None or dec_targ is None:
+        log.warning("TARG_RA and/or TARG_DEC not found; can't compute "
+                    "pixel location of target.")
+        locn = None
+    else:
+        shape = input_model.data.shape
+        grid = np.indices(shape[-2:])
+        z = np.zeros(shape[-2:], dtype=np.float64) + shape[0] // 2
+        # The arguments are the X, Y, and Z pixel coordinates, and the
+        # output arrays will be 2-D.
+        (ra_i, dec_i, wl) = input_model.meta.wcs(grid[1], grid[0], z)
+        rect = sph_to_rect(ra_i, dec_i)
+        v = sph_to_rect(ra_targ, dec_targ)
+        diff = rect - v
+        # We want the pixel with the minimum distance from v, but the pixel
+        # with the minimum value of distance squared will be the same.
+        dist2 = (diff**2).sum(axis=-1)
+        nan_mask = np.isnan(wl)
+        dist2[..., :] = np.where(nan_mask, HUGE_DIST, dist2[..., :])
+        del nan_mask
+        k = np.argmin(dist2.ravel())
+        (j, i) = divmod(k, dist2.shape[1])      # y, x coordinates
+
+        if i <= 0 or j <= 0 or i >= shape[-1] - 1 or j >= shape[-2] - 1:
+            log.warning("WCS implies the target is at or beyond the edge "
+                        "of the image; this location will not be used.")
+            locn = None
+        else:
+            locn = (i, j)                       # x, y coordinates
+
+    return locn
+
+
+def sph_to_rect(ra, dec):
+    """Convert celestial coordinates to rectangular.
```

Comment:
Cartesian?

Reply:
I thought they were the same.  I'll change this.

Reply:
That may be ...